Re-working of API to be based on downloading the full database

mattpitkin opened the pull request:
This PR is to change the PSRQpy query API, so that it defaults to downloading the whole ATNF Pulsar Catalogue database file and queries that locally. This should be more efficient that sending webform queries for each request. This will also enable to query class to load other local database files if requested.

This is currently a work in progress and requires a lot more work.

This subsumes #39.

pep8speaks commented:
Hello @mattpitkin! Thanks for submitting the PR.

 - In the file [`psrqpy/search.py`](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py), following                are the PEP8 issues :
> [Line 138:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L138): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 100 characters)
> [Line 180:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L180): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 195:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L195): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (159 > 100 characters)
> [Line 201:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L201): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 214:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L214): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (135 > 100 characters)
> [Line 291:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L291): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (141 > 100 characters)
> [Line 332:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L332): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 100 characters)
> [Line 358:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L358): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 392:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L392): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 100 characters)
> [Line 402:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L402): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (123 > 100 characters)
> [Line 437:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L437): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (135 > 100 characters)
> [Line 470:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L470): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 100 characters)
> [Line 479:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L479): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 100 characters)
> [Line 488:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L488): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 489:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L489): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)
> [Line 491:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L491): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (120 > 100 characters)
> [Line 526:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L526): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (134 > 100 characters)
> [Line 527:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L527): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (176 > 100 characters)
> [Line 534:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L534): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (145 > 100 characters)
> [Line 538:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L538): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (155 > 100 characters)
> [Line 540:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L540): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (153 > 100 characters)
> [Line 545:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L545): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (120 > 100 characters)
> [Line 550:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L550): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (135 > 100 characters)
> [Line 558:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L558): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 100 characters)
> [Line 678:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L678): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 100 characters)
> [Line 681:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L681): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 687:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L687): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (127 > 100 characters)
> [Line 701:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L701): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 100 characters)
> [Line 705:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L705): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (140 > 100 characters)
> [Line 715:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L715): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (143 > 100 characters)
> [Line 758:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L758): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 100 characters)
> [Line 774:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L774): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)
> [Line 784:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L784): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 100 characters)
> [Line 800:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L800): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 810:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L810): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 100 characters)
> [Line 856:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L856): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 100 characters)
> [Line 921:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L921): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)
> [Line 950:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L950): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 100 characters)
> [Line 953:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L953): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 955:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L955): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 956:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L956): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (156 > 100 characters)
> [Line 958:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L958): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 959:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L959): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 998:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L998): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 1000:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1000): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 1017:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1017): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 1031:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1031): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 100 characters)
> [Line 1032:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1032): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)
> [Line 1033:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1033): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 100 characters)
> [Line 1042:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1042): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 100 characters)
> [Line 1043:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1043): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 100 characters)
> [Line 1044:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1044): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 100 characters)
> [Line 1045:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1045): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (127 > 100 characters)
> [Line 1046:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1046): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (126 > 100 characters)
> [Line 1047:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1047): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (129 > 100 characters)
> [Line 1048:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1048): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (120 > 100 characters)
> [Line 1049:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1049): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (129 > 100 characters)
> [Line 1119:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1119): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 1137:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/search.py#L1137): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)


 - In the file [`psrqpy/utils.py`](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py), following                are the PEP8 issues :
> [Line 19:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L19): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 252:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L252): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)
> [Line 264:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L264): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 100 characters)
> [Line 358:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L358): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (119 > 100 characters)
> [Line 447:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L447): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 100 characters)
> [Line 454:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L454): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (136 > 100 characters)
> [Line 456:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L456): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (148 > 100 characters)
> [Line 464:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L464): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 100 characters)
> [Line 484:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L484): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (137 > 100 characters)
> [Line 495:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L495): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (155 > 100 characters)
> [Line 517:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L517): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (120 > 100 characters)
> [Line 588:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L588): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 100 characters)
> [Line 595:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L595): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (142 > 100 characters)
> [Line 597:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L597): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 100 characters)
> [Line 733:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L733): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (128 > 100 characters)
> [Line 735:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L735): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (128 > 100 characters)
> [Line 739:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L739): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 748:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L748): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 100 characters)
> [Line 753:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L753): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 761:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L761): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 769:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L769): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 100 characters)
> [Line 777:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L777): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 100 characters)
> [Line 828:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L828): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 100 characters)
> [Line 832:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L832): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 886:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L886): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)
> [Line 887:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L887): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 100 characters)
> [Line 891:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L891): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)
> [Line 895:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L895): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)
> [Line 962:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L962): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (141 > 100 characters)
> [Line 963:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L963): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (124 > 100 characters)
> [Line 964:101](https://github.com/mattpitkin/psrqpy/blob/5e645d8b8c025fdefcdb431a2817e0be8727f336/psrqpy/utils.py#L964): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)


 - Complete extra results for this file :

> file_to_check.py:820:-271: W605 invalid escape sequence '\d'
file_to_check.py:854:-333: W605 invalid escape sequence '\d'
file_to_check.py:884:-201: W605 invalid escape sequence '\!'
file_to_check.py:884:-192: W605 invalid escape sequence '\!'
file_to_check.py:884:-174: W605 invalid escape sequence '\d'
file_to_check.py:917:-336: W605 invalid escape sequence '\d'
file_to_check.py:917:-314: W605 invalid escape sequence '\l'
file_to_check.py:917:-294: W605 invalid escape sequence '\!'
file_to_check.py:917:-285: W605 invalid escape sequence '\!'
---

respiewak commented on psrqpy/utils.py:
```diff
@@ -66,20 +70,18 @@ def get_catalogue(path_to_db=None):
         # get the tarball
         try:
-            pulsargzfile = requests.get(ATNF_TARBALL)
-            fp = BytesIO(pulsargzfile.content)  # download and store in memory
```
You removed this line, but not the `fp.close` line (was 217).

mattpitkin commented on psrqpy/search.py:
```diff
@@ -179,29 +203,46 @@ def __init__(self, params=None, condition=None, psrtype=None, assoc=None, bincom
+        # check sort order is either 'asc' or 'desc' (or some synonyms)
+        if sort_order.lower() in ['asc', 'up', '^']:
+            self._sort_order = 'asc'
+        elif sort_order.lower() in ['desc', 'descending', 'v']:
```
Yes, I should really have:

```python
if sort_order.lower() in ['asc', 'ascending', 'up', '^']
    self._sort_order = 'asc'
elif sort_order.lower() in ['desc', 'descending', 'down', 'v']
```

I'll fix this.

mattpitkin commented on psrqpy/utils.py:
```diff
@@ -45,6 +47,7 @@ def get_catalogue(path_to_db=None):
         path_to_db (str): if the path to a local version of the database file
             is given then that will be read in rather than attempting to
             download the file (defaults to None).
+        cache (bool): cache the downloaded ATNF Pulsar Catalogue file.
```
I've added this to the comment string.

mattpitkin commented on psrqpy/search.py:
```diff
@@ -116,21 +140,21 @@ def __init__(self, params=None, condition=None, psrtype=None, assoc=None, bincom
+            self.load(loadquery)
             return
 
-        # check sort order is either 'asc' or 'desc' (or some synonyms)
```
Well spotted. I'll sort this out.

respiewak commented:
Let me know if you want me to do another review; otherwise I'll take the weekend off :-)

mattpitkin commented:
@respiewak thanks for looking at it so far. I expect by early next week I'll have a fully working version that can be tested, so I'll let you know when that's the case.

respiewak commented:
As of right now, the query test fails because 'JNAME' isn't in the table. Do you know why this is?

respiewak commented:
I noticed you added the derived parameters, which is awesome! 

I also noticed that the `query_test` script doesn't test very much besides whether or not the references exist. I tried adding a `getitem` for each parameter and found that it breaks when it reaches 'GL', so there's still something to do there.

respiewak commented:
With the current state of things, it appears this could nearly replace the `psrcat` software (although still requiring the catalogue file). Do you intend to continue in this manner and implement command line parsing (e.g., with `argparse`)?

mattpitkin commented:
> I noticed you added the derived parameters, which is awesome!

Yeah, it made the change a bit more complicated than I'd initially imagined, but it's hopefully worth it and allowed me to completely drop the need to try and generate a query through the catalogue web form.

> I also noticed that the query_test script doesn't test very much besides whether or not the references exist. I tried adding a getitem for each parameter and found that it breaks when it reaches 'GL', so there's still something to do there.

I've not quite finished adding all the derived parameters that are in `psrcat` yet, but there aren't too many more to go. The `query_test` script does really need changing though. I should probably have a [`unittest`](https://docs.python.org/2/library/unittest.html) script instead, and and test database file that contains a set of known pulsar parameters (maybe taken from the ATNF catalogue) that can be tested against in the test script (also with known values of the derived parameters). This will take a bit more work, but shouldn't be a problem.

mattpitkin commented:
> With the current state of things, it appears this could nearly replace the psrcat software (although still requiring the catalogue file). Do you intend to continue in this manner and implement command line parsing (e.g., with argparse)?

Having a `psrcat`-like command line interface sound like a good idea, and relatively simple thing to do. One thing I planned to do after this PR was actually move PSRQpy into [astroquery](https://astroquery.readthedocs.io/en/latest/) (see astropy/astroquery#1227) and continue development there. That might not allow for a command line interface (as far as I can tell astroquery doesn't install any runnable scripts, but maybe they would consider it), but it could be that PSRQpy could then just become a `psrcat`-like command line interface for the astroquery code.

respiewak commented:
Besides the conflicts, I'm now getting a `ValueError` when running the (slightly edited) `query_test` script: ```ValueError: Velocity data is currently only supported in the coordinate frame objects, not in SkyCoord. See the velocities documentation page for more information: http://docs.astropy.org/en/stable/coordinates/velocities.html```
Is a specific version of `astropy` required for these new changes?

respiewak commented:
Will the PEP8 issues never end?!? 

What else needs to be added to this PR before merging? I'll make some more changes to the parsing functions (including proper docstrings)...

respiewak commented on test/query_test.py:
```diff
@@ -90,6 +90,21 @@ def test_derived_p0(query_derived, query_atnf):
+    elong = query_derived.get_pulsar('TEST1')['ELONG'][0]
+    elongatnf = query_atnf.get_pulsar('TEST1')['ELONG'][0]
+
+    assert abs(elong-elongatnf) < 1e-5
```
This is a good test, but why 1e-5? What is the usual precision of the `elong` and `elat` values in the catalogue?

mattpitkin replied:
The `1e-5` was just put in because it worked (`1e-6` fails). I'm not actually sure what consistency to expect between how `psrcat` calculates the ecliptic coordinates and how `psrqpy` does it using astropy. It might be that we're being more accurate, or they're both correct, but using slightly different conventions somewhere.

respiewak commented on .travis.yml:
```diff
@@ -16,21 +17,18 @@ install:
   # install packages for documentation building
   - pip install sphinx sphinx-rtd-theme
+  # install packages for testing
+  - pip install pytest pytest-socket
```
Should this be `pytest_socket`?

respiewak replied:
Hmm, both seem to work... weird

respiewak commented on psrqpy/search.py:
```diff
@@ -18,586 +18,2043 @@
+            idx = np.isfinite(A1) & np.isfinite(PB)
+
+            MASSFN = np.full(self.catalogue_len, np.nan)
+            MASSFN[idx] = (4.*np.pi**2/GM_sun.value)*A1[idx]**3/(PB[idx]**2)
```
I think you might be missing a few factors of c here. The equation should be `f = 4*np.pi**2 * A1[idx]**3 * c**3 / (GM_sun * PB[idx]**2)`

mattpitkin replied:
> I think you might be missing a few factors of c here. The equation should be `f = (4*np.pi**2 * A1[idx]**3 * c**3) / (GM_sun.value * PB[idx]**2)`

The `A1` values have already been multiplied by c a few lines above.

mattpitkin replied:
> Also, it might be good to add the error in the mass function. It's not included in the psrcat, but it's useful.

Yes, I would be useful to have. Do you agree that the error should be:

```python
A1ERR = self.catalogue['A1_ERR']*c.value
PBERR = self.catalogue['PB_ERR']*86400.

idx = np.isfinite(MASSFN) & np.isfinite(A1ERR) & np.isfinite(PBERR)

MASSFN_ERR = np.full(self.catalogue_len, np.nan)
MASSFN_ERR[idx] = MASSFN[idx] * np.sqrt( (3.*A1ERR[idx]/A1[idx])**2 + (2.*PBERR[idx]/PB[idx])**2 )
```
I just used standard propagation of errors (assuming `A1` and `PB` are uncorrelated), but it would be good if you could double check.

respiewak replied:
> > I think you might be missing a few factors of c here. The equation should be `f = (4*np.pi**2 * A1[idx]**3 * c**3) / (GM_sun.value * PB[idx]**2)`
> 
> The `A1` values have already been multiplied by c a few lines above.

OK, I didn't see that. It might be good to add a comment about that change of units, though

respiewak replied:
> > Also, it might be good to add the error in the mass function. It's not included in the psrcat, but it's useful.
> 
> Yes, I would be useful to have. Do you agree that the error should be:
> 
> ```python
> A1ERR = self.catalogue['A1_ERR']*c.value
> PBERR = self.catalogue['PB_ERR']*86400.
> 
> idx = np.isfinite(MASSFN) & np.isfinite(A1ERR) & np.isfinite(PBERR)
> 
> MASSFN_ERR = np.full(self.catalogue_len, np.nan)
> MASSFN_ERR[idx] = MASSFN[idx] * np.sqrt( (3.*A1ERR[idx]/A1[idx])**2 + (2.*PBERR[idx]/PB[idx])**2 )
> ```
> 
> I just used standard propagation of errors (assuming `A1` and `PB` are uncorrelated), but it would be good if you could double check.

I think that looks right!

mattpitkin replied:
Cool, I'll add this in now.

respiewak commented on psrqpy/search.py:
```diff
@@ -18,586 +18,2043 @@
+            EPS1 = self.catalogue['EPS1'].values.copy()
+            EPS2 = self.catalogue['EPS2'].values.copy()
+            ECCnew = np.full(self.catalogue_len, np.nan)
+            OMnew = np.full(self.catalogue_len, np.nan)
```
This section appears to assume that `ECC` and `OM` never exist and must always be derived from `EPS1` and `EPS2`, but that's not the case, as far as I know. Occasionally, people will publish `OM` and `ECC` values, so this calculation should go both ways.

mattpitkin replied:
Yes, I will change things, so that `EPS1` and `EPS2` also get calculated.

mattpitkin replied:
I've added calculations for `EPS1`, `EPS2` and their respective errors now.

mattpitkin commented on psrqpy/search.py:
```diff
@@ -18,586 +18,2043 @@
+            SINI_MIN = 1.0  # inclination for minimum mass
+            SINI_MED = 0.866025403  # inclination of 60deg for median mass
+            SINI_90 = 0.438371146   # inclination for 90% UL mass
+            for i, mf in enumerate(MASSFN):
```
I've looked at a couple of examples and they were consistent with what is in the ATNF and produced by `psrcat`, although in general the masses all looked rather low. It may be that what's currently done in `psrcat`, which I've copied, needs to be updated. So, we could keep the values we currently have for consistency, but if there's another way of doing things we could also add that into `psrqpy`.

respiewak replied:
I might take a more careful look there, and see what's done in the literature for better reference.

respiewak replied:
So I think I was misunderstanding what `newton` actually does. It doesn't allow `m_p` to vary, correct? Just `m_c`? That should all be fine, and I haven't found anything different in the literature.

respiewak commented on psrqpy/search.py:
```diff
@@ -1449,6 +1449,48 @@ def solfunc(m2, sini, mf, m1):
+
+            idx = np.isfinite(ECC) & np.isfinite(OM)
+
+            EPS1new[idx] = ECC[idx] * np.sin(OM[idx])
```
If you use `EPS1new = np.full(self.catalogue_len, np.nan)` and then do the calculation in this way, don't you overwrite the EPS1 (and EPS2) values that were previously in the catalogue but not associated with ECC/OM values (and vice versa with the previous calculations)? Or am I misunderstanding how `self.update` works? 

I would think you'd want to do: 
```if 'EPS1' in self.columns:
    EPS1new = self.catalogue['EPS1'].values.copy()
else:
    EPS1new = np.full(self.catalogue_len, np.nan)```
and so on for EPS2, and then for ECC and OM in the previous derivation.

mattpitkin replied:
By default `self.update` is set to only update non-NaN values if they already exist in the table, so it shouldn't overwrite existing values.

respiewak commented on test/test_catalogue.db:
```diff
@@ -4,6 +4,7 @@
 # TEST1 is based on J0023+0923
 # TEST2 is based on J0026+6320
 # TEST3 is based on J0357+3205
+# TEST4 is based on J0024-7204J
```
Pretty sure this should be "J0024-7204O" as in `derived_catalogue.db`

mattpitkin replied:
Well spotted. I've fixed this now.

mattpitkin commented:
Hi @respiewak, I think I've got to the position where I'm happy with the current changes and think it would be good to get this merged into master. If you think it's ok too I'd be happy for you to go ahead and press the merge button!

When it's merged I'll make a new release and push it to PyPI.

respiewak commented:
Nice! I'll give it a read-through to make sure nothing's missing. Is the README up-to-date? It looks like the requirements list is missing pandas still... And do the tests run automatically during installation, or should a comment be added to the README about running them?

respiewak commented:
Also, is it just me, or is a `QueryATNF` object from the full catalogue very slow to work with? Just doing `q.dataframe` took more than 20 seconds. This might be something that can wait for the next release, but it should definitely be on the list of things to look into.

respiewak commented on psrqpy/search.py:
```diff
@@ -15,686 +15,2378 @@
+        TYPEnew[idxtype] = TYPEnew[idxtype].apply(lambda x: re.split(r'\[|,|\(|:', x)[0])
+
+        self.update(TYPEnew)
+        self.update(TYPEREFnew)
```
Oh, I think I figured it out:
```suggestion
        self.update(TYPEREFnew, overwrite=True)
```
should do the trick. Actually, I'm not sure why this is required for the 'TYPE_REF' column, which should only have NaNs before this line, but it definitely is required for the `self.update(TYPEnew)` line

mattpitkin replied:
Thanks for spotting this, I think I've fixed it in fc7efac

respiewak replied:
Hmm, I think that's not quite right. The problem appears to be that `update` ignores all values that are not NaNs (in the original column) when adding new values, so it doesn't change, for example, "HE[aaa+18]" into "HE", so we need to add `overwrite=True` to the `update` lines.

respiewak replied:
Do you think this problem exists in any other functions? Are there any that would be updating an existing value instead of adding a new (non-NaN) value? ... I could probably check myself...

mattpitkin replied:
Thanks for the fix. I can't think of anywhere else this would be a problem.

mattpitkin commented:
> Nice! I'll give it a read-through to make sure nothing's missing. Is the README up-to-date? It looks like the requirements list is missing pandas still... And do the tests run automatically during installation, or should a comment be added to the README about running them?

I've updated the README to give the proper requirements including pandas. I've also added a note about running the test suite - this is not run during installation, but would have to be run by the user using pytest.

mattpitkin commented:
>> Also, is it just me, or is a QueryATNF object from the full catalogue very slow to work with? Just doing q.dataframe took more than 20 seconds. This might be something that can wait for the next release, but it should definitely be on the list of things to look into.

> Yeah, it's pretty slow to create the table if you use the include_refs=True flag, otherwise it's nice an quick. This is because each time it tries to create a dataframe/table based on the user selections, including whether or not to convert the short reference strings into the full reference. If you are just interested in a few pulsars and have explicitly asked for them by name then it's still quite quick, as it only has to do the reference conversion for those few. I think this could definitely be sped up and made more efficient in a future release.

I've just changed things a bit here. Now, if you specify `include_refs` it will just return the short ATNF reference tags, rather than trying to convert them to the full reference string. This means that returning the table should stay nice and quick. If you actually want to set what the actual reference for a given short tag is you can now pass it to the `parse_ref` method of the `QueryATNF` class and it will give you back the full reference string. I think this is probably all that's really needed.

respiewak commented:
Awesome! It looks like it's ready for merging then!

respiewak commented on docs/source/index.rst:
```diff
@@ -238,8 +238,8 @@ We can make use of the `ADS module <https://ads.readthedocs.io/en/latest/>`_ to
     >>> from psrqpy import QueryATNF
-    >>> query = QueryATNF(params='PB', psrs='J0737-3039A', include_refs=True, adsref=True)
-    >>> print(query.table['PB_REFURL'][0])
+    >>> query = QueryATNF(params='PB', psrs='J0737-3039A', adsref=True)
```
Doesn't this still need the `include_refs` argument? It doesn't appear to work right now

mattpitkin replied:
Ah, yes. I briefly removed the `include_refs` options during testing, but then put it back in without re-fixing the docs. I've done that now in c00b4f96f3b

respiewak commented:
Merge! Before we find some other little thing to mess with :-)


On Fri, 16 Nov 2018, 9:01 PM Matt Pitkin <notifications@github.com> wrote:

> *@mattpitkin* commented on this pull request.
> ------------------------------
>
> In docs/source/index.rst
> <https://github.com/mattpitkin/psrqpy/pull/41#discussion_r234149062>:
>
> > @@ -238,8 +238,8 @@ We can make use of the `ADS module <https://ads.readthedocs.io/en/latest/>`_ to
>  For example we could get the reference for the orbital period of J0737-3039A with
>
>      >>> from psrqpy import QueryATNF
> -    >>> query = QueryATNF(params='PB', psrs='J0737-3039A', include_refs=True, adsref=True)
> -    >>> print(query.table['PB_REFURL'][0])
> +    >>> query = QueryATNF(params='PB', psrs='J0737-3039A', adsref=True)
>
> Ah, yes. I briefly removed the include_refs options during testing, but
> then put it back in without re-fixing the docs. I've done that now in
> c00b4f9
> <https://github.com/mattpitkin/psrqpy/commit/c00b4f96f3b21b65374ad872fd56010ca075ab8c>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mattpitkin/psrqpy/pull/41#discussion_r234149062>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AaH-8MXLBvFl0naYHFDAoXk5Iux6QDOxks5uvozygaJpZM4WoQXY>
> .
>

mattpitkin commented:
> Merge! Before we find some other little thing to mess with :-)

I've already started messing ;-)

mattpitkin commented:
Merged! :tada:

respiewak commented:
Hooray! Great work on this!

On Fri, 16 Nov 2018, 9:32 PM Matt Pitkin <notifications@github.com> wrote:

> Merged! 🎉
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mattpitkin/psrqpy/pull/41#issuecomment-439351769>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AaH-8ORXNryb0oeDs0gosFxYJfVOrECCks5uvpQ5gaJpZM4WoQXY>
> .
>

mattpitkin commented:
Thanks for all your help and spotting many issues. I'll create a new PyPI release now too.
